lisa/tests/staging/schedutil.py

```diff
@@ -0,0 +1,271 @@
+            (df['util'] > df['util_est_enqueued'])
+        )
+        df['boost_points'] = boost_points
+
+        df['expected_cost_margin'] = (df['util'] - df['util_est_enqueued']).where(
+            cond=boost_points,
+            other=0,
+        )
+
+        # cost_margin values range from 0 to 1024
+        ENERGY_SCALE = 1024
+
+        for col in ('expected_cost_margin', 'cost_margin'):
+            df[col] *= 100 / ENERGY_SCALE
+
+        # We cannot know if the first row is supposed to be boosted or not
+        # because we lack history, so we just drop it
+        return df.iloc[1:]
+
+    def _plot_test_boost(self, df):
+        task = self.rtapp_tasks[0]
+        analysis = self.trace.analysis.frequency
+        fig, axis = analysis.setup_plot(interactive=False)
+        df['cost_margin'].plot(ax=axis, drawstyle='steps-post', color='r')
+        df['boost_points'].astype('int', copy=False).plot(ax=axis, drawstyle='steps-post', color='black')
+        df['expected_cost_margin'].plot(ax=axis, drawstyle='steps-post', color='blue')
+
+        self.trace.analysis.tasks.plot_task_activation(task, axis=axis, overlay=True)
+
+        axis.legend()
+        analysis.save_plot(fig, filepath=os.path.join(self.res_dir, 'ramp_boost.svg'))
+        return axis
+
+    @requires_events('sched_switch')
+    def trace_window(self, trace):
+        """
+        Skip the first phase in the trace, since it is heavily influenced by
+        what was immediately preceding it.
+        """
+        profile = self.rtapp_profile
+        phase_duration = max(
+            task.phases[0].duration_s
+            for task in profile.values()
+        )
+        start, end = super().trace_window(trace)
+        return (start + phase_duration, end)
+
+    @df_ramp_boost.used_events
+    def test_ramp_boost(self, nrg_threshold_pct=0.1, bad_samples_threshold_pct=0.1) -> ResultBundle:
+        """
+        Test that the energy boost feature is triggering as expected.
+        """
+        # If there was no cost_margin sample to look at, that means boosting
+        # was not exhibited by that test so we cannot conclude anything
+        df = self.df_ramp_boost()
+        self._plot_test_boost(df)
+
+        if df.empty:
+            return ResultBundle(Result.UNDECIDED)
+
+        # Make sure the boost is always positive (negative cannot really happen
+        # since the kernel is using unsigned arithmetic, but still check in
+        # case there are some dataframe handling issues)
+        assert not (df['expected_cost_margin'] < 0).any()
+        assert not (df['cost_margin'] < 0).any()
+
+        # "rect" method is accurate here since the signal is really following
+        # "post" steps
+        expected_boost_nrg = series_mean(df['expected_cost_margin'])
+        actual_boost_nrg = series_mean(df['cost_margin'])
+
+        # Check that the total amount of boost is close to expectations
+        lower = max(0, expected_boost_nrg - nrg_threshold_pct)
```

Comment:
Shouldn't that `nrg_threshold_pct` be multiplied to something?

Reply:
No, these series are already in percents